staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -238,43 +306,242 @@ func newAuthenticator(opts Options, initVerifier func(ctx context.Context, a *Au
 		Now:                  now,
 	}
 
+	// TODO: Distributed claims resolution will be enabled unconditionally once
+	// response caching is implemented.
+	var resolver *claimResolver
+	if opts.enableDistributedClaims {
+		resolver = newClaimResolver(opts.GroupsClaim, client, verifierConfig)
+	}
+
+	authenticator := &Authenticator{
+		issuerURL:      opts.IssuerURL,
+		usernameClaim:  opts.UsernameClaim,
+		usernamePrefix: opts.UsernamePrefix,
+		groupsClaim:    opts.GroupsClaim,
+		groupsPrefix:   opts.GroupsPrefix,
+		requiredClaims: opts.RequiredClaims,
+		cancel:         cancel,
+		resolver:       resolver,
+	}
+
 	initVerifier(ctx, authenticator, verifierConfig)
 	return authenticator, nil
 }
 
-func hasCorrectIssuer(iss, tokenData string) bool {
-	parts := strings.Split(tokenData, ".")
+// untrustedIssuer extracts an untrusted "iss" claim from the given JWT token,
+// or returns an error if the token can not be parsed.  Since the JWT is not
+// verified, the returned issuer should not be trusted.
+func untrustedIssuer(token string) (string, error) {
+	parts := strings.Split(token, ".")
 	if len(parts) != 3 {
-		return false
+		return "", fmt.Errorf("malformed token %q", token)
 	}
 	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
 	if err != nil {
-		return false
+		return "", fmt.Errorf("error decoding token %q: %v", token, err)
 	}
 	claims := struct {
 		// WARNING: this JWT is not verified. Do not trust these claims.
 		Issuer string `json:"iss"`
 	}{}
 	if err := json.Unmarshal(payload, &claims); err != nil {
+		return "", fmt.Errorf("while unmarshaling token: %q: %v", err)
+	}
+	return claims.Issuer, nil
+}
+
+func hasCorrectIssuer(iss, tokenData string) bool {
+	uiss, err := untrustedIssuer(tokenData)
+	if err != nil {
+		glog.Errorf("while checking issuer: %v", err)
```

Comment:
Please drop this. It's going to be called for every non-oidc token that goes through the oidc authenticator.